components/encryption/src/master_key/kms.rs

```diff
@@ -0,0 +1,306 @@
+            encryption_context: None,
+            grant_tokens: None,
+            encryption_algorithm: None,
+            key_id: None,
+        };
+        let decrypt_response = retry(runtime, || {
+            self.client
+                .decrypt(decrypt_request.clone())
+                .map_err(|e| Error::Other(e.into()))
+        });
+        let plaintext = decrypt_response.plaintext.unwrap().as_ref().to_vec();
+        Ok(plaintext)
+    }
+
+    fn generate_data_key(&self, runtime: &mut Runtime) -> Result<(Vec<u8>, Vec<u8>)> {
+        let generate_request = GenerateDataKeyRequest {
+            encryption_context: None,
+            grant_tokens: None,
+            key_id: self.current_key_id.clone(),
+            key_spec: Some(AWS_KMS_DATA_KEY_SPEC.to_owned()),
+            number_of_bytes: None,
+        };
+        let generate_response = retry(runtime, || {
+            self.client
+                .generate_data_key(generate_request.clone())
+                .map_err(|e| Error::Other(e.into()))
+        });
+        let ciphertext_key = generate_response.ciphertext_blob.unwrap().as_ref().to_vec();
+        let plaintext_key = generate_response.plaintext.unwrap().as_ref().to_vec();
+        Ok((ciphertext_key, plaintext_key))
+    }
+}
+
+fn retry<T, U, F>(runtime: &mut Runtime, mut func: F) -> T
+where
+    F: FnMut() -> U,
+    U: Future<Output = Result<T>> + std::marker::Unpin,
+{
+    let retry_limit = 6;
+    let timeout_duration = Duration::from_secs(10);
+    for _ in 0..retry_limit {
+        let fut = func();
+
+        match runtime.block_on(async move {
+            let timeout = tokio::time::delay_for(timeout_duration);
+            future::select(fut, timeout).await
+        }) {
+            future::Either::Left((Ok(resp), _)) => return resp,
+            future::Either::Left((Err(e), _)) => {
+                error!("kms request failed"; "error"=>?e);
+            }
+            future::Either::Right((_, _)) => {
+                error!("kms request timeout"; "timeout" => ?timeout_duration);
+            }
+        }
+    }
+    panic!("kms request failed in {} times", retry_limit)
+}
+
+pub struct KmsBackend {
+    file_backend: FileBackend,
+}
+
+impl KmsBackend {
+    pub fn new(config: KmsConfig, base: &str) -> Result<KmsBackend> {
+        let kms = AwsKms::new(config)?;
+        KmsBackend::with_kms(kms, base)
+    }
+
+    fn with_kms(kms: AwsKms, base: &str) -> Result<KmsBackend> {
+        let vendor_backend = WithMetadata {
+            // For now, we only support AWS.
+            metadata: vec![(MetadataKey::KmsVendor, AWS_KMS_VENDOR_NAME.to_vec())],
+            // ciphertext_key has already be ecrypted by KMS.
+            backend: PlainTextBackend::default(),
+        };
+
+        // Create base dir if it is missing.
+        create_dir_all(base)?;
+
+        // Read the master key or generate a new master key.
+        let key_path = Path::new(base).join(KMS_ENCRYPTION_KEY_NAME);
```

Comment:
I think there's no need to create a new file to store the key. We can store it in metadata of `EncryptedContent`.